From fda9fa19b09067b6a3ee6ff8d93e977957e0655c Mon Sep 17 00:00:00 2001
From: Chris Wilson <chris@chris-wilson.co.uk>
Date: Thu, 12 Sep 2019 17:08:34 +0100
Subject: drm/i915: Don't mix srcu tag and negative error codes
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit
Git-commit: fda9fa19b09067b6a3ee6ff8d93e977957e0655c
Patch-mainline: v5.4-rc3
References: bsc#1152489

While srcu may use an integer tag, it does not exclude potential error
codes and so may overlap with our own use of -EINTR. Use a separate
outparam to store the tag, and report the error code separately.

Fixes: 2caffbf11762 ("drm/i915: Revoke mmaps and prevent access to fence registers across reset")
Signed-off-by: Chris Wilson <chris@chris-wilson.co.uk>
Cc: Mika Kuoppala <mika.kuoppala@linux.intel.com>
Cc: Ville Syrjälä <ville.syrjala@linux.intel.com>
Reviewed-by: Mika Kuoppala <mika.kuoppala@linux.intel.com>
Link: https://patchwork.freedesktop.org/patch/msgid/20190912160834.30601-1-chris@chris-wilson.co.uk
(cherry picked from commit eebab60f224fcfd560957715d08c31564d8672ed)
Signed-off-by: Rodrigo Vivi <rodrigo.vivi@intel.com>
Acked-by: Thomas Zimmermann <tzimmermann@suse.de>
---
 drivers/gpu/drm/i915/gem/i915_gem_mman.c |    6 ++----
 drivers/gpu/drm/i915/gt/intel_reset.c    |    7 +++----
 drivers/gpu/drm/i915/gt/intel_reset.h    |    2 +-
 3 files changed, 6 insertions(+), 9 deletions(-)

--- a/drivers/gpu/drm/i915/gem/i915_gem_mman.c
+++ b/drivers/gpu/drm/i915/gem/i915_gem_mman.c
@@ -246,11 +246,9 @@ vm_fault_t i915_gem_fault(struct vm_faul
 
 	wakeref = intel_runtime_pm_get(rpm);
 
-	srcu = i915_reset_trylock(i915);
-	if (srcu < 0) {
-		ret = srcu;
+	ret = i915_reset_trylock(i915, &srcu);
+	if (ret)
 		goto err_rpm;
-	}
 
 	ret = i915_mutex_lock_interruptible(dev);
 	if (ret)
--- a/drivers/gpu/drm/i915/gt/intel_reset.c
+++ b/drivers/gpu/drm/i915/gt/intel_reset.c
@@ -1404,10 +1404,9 @@ out:
 	intel_runtime_pm_put(&i915->runtime_pm, wakeref);
 }
 
-int i915_reset_trylock(struct drm_i915_private *i915)
+int i915_reset_trylock(struct drm_i915_private *i915, int *srcu)
 {
 	struct i915_gpu_error *error = &i915->gpu_error;
-	int srcu;
 
 	might_lock(&error->reset_backoff_srcu);
 	might_sleep();
@@ -1423,10 +1422,10 @@ int i915_reset_trylock(struct drm_i915_p
 
 		rcu_read_lock();
 	}
-	srcu = srcu_read_lock(&error->reset_backoff_srcu);
+	*srcu = srcu_read_lock(&error->reset_backoff_srcu);
 	rcu_read_unlock();
 
-	return srcu;
+	return 0;
 }
 
 void i915_reset_unlock(struct drm_i915_private *i915, int tag)
--- a/drivers/gpu/drm/i915/gt/intel_reset.h
+++ b/drivers/gpu/drm/i915/gt/intel_reset.h
@@ -35,7 +35,7 @@ int i915_reset_engine(struct intel_engin
 
 void i915_reset_request(struct i915_request *rq, bool guilty);
 
-int __must_check i915_reset_trylock(struct drm_i915_private *i915);
+int __must_check i915_reset_trylock(struct drm_i915_private *i915, int *srcu);
 void i915_reset_unlock(struct drm_i915_private *i915, int tag);
 
 int i915_terminally_wedged(struct drm_i915_private *i915);
